src/algebra/roption_instances.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2019 Hoang Le Truong. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author : Hoang Le Truong.
+
+roption α  can see as  Prop × α. (Prop, ∧) has natural algebra structure. Therefore if α is a semigroup, a comm_semigroup, 
```

Comment:
> roption α  can see as  Prop × α

No. This makes no sense.

Reply:
> (Prop, ∧) has natural algebra structure.

What do you mean? I agree that `(Prop, ∧)` is a semigroup, and I won't complain if you say it's a monoid, but I'm not sure what you mean otherwise. I don't understand the "Therefore" in the next sentence at all.

Reply:
Since ```get := dom → α``` is map with ```dom```, algebraic structure of ```dom```  with operation ```∧``` will compatible with  algebraic structure of α. Now we can forget ```get``` is map and see ```get``` as ```α```.  Because ```roption α := ⟨dom,get⟩```  algebraic structure of  ```roption α``` is dependent on algebraic structure of ```dom``` and ```get``` as ```dom × get```. Notice that ```(Prop,∧)``` is semigroup. Therefore algebraic structure of  ```roption α``` is  only dependent on  algebraic structure of   ```get```. Hence if ```α``` is semigroup, so it ```roption α```.

Reply:
Suppose that `α` has 2 elements (e.g. it is `fin 2`). Then `Prop × α` has 4 elements (at least classicaly I can prove that). However, `roption α` has 3 elements. So something is wrong here...

Reply:
I used ```forget``` is map ```roption α → dom × α```.

Reply:
`roption α` is called that because it's classically `option α`, i.e. `α + 1`, not ` Prop × α` which is classically `2 × α`. There is an algebraic structure on `roption α` but only subject to the restriction that all operations are strict, i.e. `none + x = none`,  `none * x = none`,  `-none = none` etc. So the only way it can be a monoid or add_monoid is if `none` is if the structure permits such an absorbing element. Many of these constructions are already worked out for `option` - see `with_top`, `with_zero`, `with_one` etc. The constraints on roption mean that the only mathlib algebraic structure it can only support is the (comm)monoid structure given a (comm)semigroup on alpha.

Reply:
@truonghoangle There is no such map. You can't forget the fact that the second component is a function, because it might be an empty function. In the case when `dom` is false, the right component is the empty function, and there is no way to extract an element of alpha.

Reply:
I am mistake. Thank you for your comments.